string finite_state_machine